pandas/tests/window/test_groupby.py

```diff
@@ -556,23 +556,31 @@ def test_groupby_rolling_nans_in_index(self, rollings, key):
         with pytest.raises(ValueError, match=f"{key} must be monotonic"):
             df.groupby("c").rolling("60min", **rollings)
 
-    def test_groupby_rolling_group_keys(self):
+    @pytest.mark.parametrize("group_keys", [True, False])
+    def test_groupby_rolling_group_keys(self, group_keys):
         # GH 37641
+        # GH 38523: GH 37641 actually was not a bug.
+        # group_keys only applies to groupby.apply directly
         arrays = [["val1", "val1", "val2"], ["val1", "val1", "val2"]]
         index = MultiIndex.from_arrays(arrays, names=("idx1", "idx2"))
 
         s = Series([1, 2, 3], index=index)
-        result = s.groupby(["idx1", "idx2"], group_keys=False).rolling(1).mean()
+        result = s.groupby(["idx1", "idx2"], group_keys=group_keys).rolling(1).mean()
         expected = Series(
             [1.0, 2.0, 3.0],
             index=MultiIndex.from_tuples(
-                [("val1", "val1"), ("val1", "val1"), ("val2", "val2")],
-                names=["idx1", "idx2"],
+                [
+                    ("val1", "val1", "val1", "val1"),
+                    ("val1", "val1", "val1", "val1"),
+                    ("val2", "val2", "val2", "val2"),
+                ],
+                names=["idx1", "idx2", "idx1", "idx2"],
```

Comment:
Are we sure this is the behaviour we want? 
It might be the most consistent, but it's also kind of useless to repeat those index levels .. So we should at least have a way to avoid getting that?

Reply:
I would say the consistency is more maintainable on our end compared to additionally including logic to de-duplicate index levels given some condition. 

I would prefer the user explicitly call `droplevel`.

Reply:
hmm
```
In [134]: df.groupby(level=0).transform('max')                                                                                                          
Out[134]: 
                Max Speed
Animal Type              
Falcon Captive      390.0
       Wild         390.0
Parrot Captive       30.0
       Wild          30.0
```

so i think we are doing some magic in groupby for this. These are conceptually similar operations.

Reply:
Before indexers were implemented for `groupby().rolling()`, this was the result:

```
In [1]: import pandas as pd

In [2]: pd.__version__
Out[2]: '1.0.5'

In [3]: from pandas import *

In [4]:         arrays = [
   ...:             ["Falcon", "Falcon", "Parrot", "Parrot"],
   ...:             ["Captive", "Wild", "Captive", "Wild"],
   ...:         ]
   ...:         index = MultiIndex.from_arrays(arrays, names=("Animal", "Type"))
   ...:         df = DataFrame({"Max Speed": [390.0, 350.0, 30.0, 20.0]}, index=index)
   ...:         result = df.groupby(level=0)["Max Speed"].rolling(2).sum()

In [5]: result
Out[5]:
Animal  Animal  Type
Falcon  Falcon  Captive      NaN
                Wild       740.0
Parrot  Parrot  Captive      NaN
                Wild        50.0
Name: Max Speed, dtype: float64
```

which I think we should be trying to match. Though I'm not sure if we have solid conventions of the resulting index when using groupby.

Reply:
yep i see that. ok i think that we *should* revert for 1.2.x and then decide for 1.3 is prob ok. i am leaning towards have the same as groupby here.

Reply:
Okay once this is merged in I can create another issue to discuss what the index behavior should be for groupby rolling with duplicate index levels.

Reply:
When comparing to 1.0.5 behaviour, we also had:

```
In [5]: s.groupby(["idx1", "idx2"], group_keys=False).rolling(1).mean() 
Out[5]: 
idx1  idx2
val1  val1    1.0
      val1    2.0
val2  val2    3.0
dtype: float64

In [9]: pd.__version__   
Out[9]: '1.0.5'
```

So this PR then deviates from that for this case.

(I know the influence of `group_keys=False` has been considered a bug, but we could also reconsider that, since the above seems to actually give the desired result?)

Reply:
The `group_keys` result different I think is due to the implementation change in `groupby().rolling()`

Before `groupby().rolling()` under the hood was `groupby().apply(lambda x: x.rolling()...)` and therefore `group_keys` impacted the result (since the argument is only applicable for `groupby().apply()`).

After `groupby().rolling()` moved away from using `apply`, `group_keys` didn't impact the result.

So IMO, `group_keys` shouldn't have ever really influenced the result since `groupby().apply()` was never called directly from the user.

Reply:
can you add testing for group_keys in any event?

Reply:
Yes that's parameterized here already: https://github.com/pandas-dev/pandas/pull/38737/files#diff-e338c43cbd06b849f3d6a1b97cd787a48770c616b627f33eb20f67a6fc56b116R559